File: misc/dfs_a.py
# https://atc001.contest.atcoder.jp/tasks/dfs_a


H, W = map(int, input().split())
grid = [input().strip() for _ in range(H)]

for i in range(H):
    for k in range(W):
        if grid[i][k] == "s":
            start = (i, k)
        if grid[i][k] == "g":
            goal = (i, k)

stack = [start]
seen = {start}
while stack:
    (i, k) = stack.pop()
    if (i, k) == goal:
        print("Yes")
        break
    for (next_i, next_k) in [(i + 1, k), (i - 1, k), (i, k - 1), (i, k + 1)]:
        criteria = (
            0 <= next_i < H
            and 0 <= next_k < W
            and (next_i, next_k) not in seen
            and (grid[next_i][next_k] == "." or grid[next_i][next_k] == "g")
        )
        if criteria:
            seen.add((next_i, next_k))
            stack.append((next_i, next_k))
else:
    print("No")

File: misc/dfs_a.cc
// https://atc001.contest.atcoder.jp/tasks/dfs_a
#include <iostream>
#include <stack>
using namespace std;

typedef pair<int, int> coord;

int H, W;
stack<coord> stk;
bool seen[501][501];
char grid[501][501];
coord start, goal;
int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};

int main()
{
    cin >> H >> W;
    for (int i = 0; i < H; ++i)
    {
        for (int j = 0; j < W; ++j)
        {
            char tmp;
            cin >> tmp;
            grid[i][j] = tmp;
            seen[i][j] = false;
            if (tmp == 's')
                start = make_pair(i, j);
            else if (tmp == 'g')
                goal = make_pair(i, j);
        }
    }

    bool flg = false;
    stk.push(start);
    seen[start.first][start.second] = true;
    while (!stk.empty())
    {
        auto state = stk.top();
        stk.pop();
        if (state == goal)
        {
            flg = true;
            break;
        }
        for (int i = 0; i < 4; ++i)
        {
            int r = state.first + dx[i];
            int c = state.second + dy[i];
            if (0 <= r && r < H && 0 <= c && c < W && grid[r][c] != '#' && !seen[r][c])
            {
                seen[r][c] = true;
                stk.emplace(make_pair(r, c));
            }
        }
    }
    string ans = flg ? "Yes" : "No";
    cout << ans << endl;
}
